File: backend/chat-service/src/main/java/com/example/chatservice/chat/dto/ChatRoomDto.java
package com.example.chatservice.chat.dto;

import com.example.chatservice.member.dto.MemberDto;
import com.example.chatservice.chat.entity.ChatRoom;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ChatRoomDto implements Serializable{
    private Long id;
    private String roomId;
    private String roomName;
    private String roomImage;
    private String lastChatMsg;
    private String lastChatId;
    private String lastChatTime;
    private List<MemberDto> members = new ArrayList<>();

    public ChatRoomDto(ChatRoom chatRoom, List<MemberDto> members) {
        setRoomId(chatRoom.getRoomId());
        setRoomName(chatRoom.getRoomName());
        setRoomImage(chatRoom.getRoomImage());
        setLastChatMsg(chatRoom.getLastChatMsg());
        setLastChatId(chatRoom.getLastChatId());
        setLastChatTime(chatRoom.getLastChatTime());
        setMembers(members);
    }

    public void updateLastChat(String chatId, ChatDto chatDto) {
        setLastChatId(chatId);
        setLastChatMsg((chatDto.getChatType() == ChatType.TEXT.getChatType()) ?
                chatDto.getMessage() : ChatType.fromChatType(chatDto.getChatType()).getChatTypeStr());
        setLastChatTime(chatDto.getChatTime());
    }

    public boolean checkChatRoomMember(String userId) {
        List<MemberDto> chatRoomMembers = members.stream()
                .filter(memberDto -> memberDto.getUserId().equals(userId))
                .collect(Collectors.toList());

        return chatRoomMembers.size() == 0;
    }
}

File: backend/chat-service/src/main/kotlin/com/example/chatservice/chat/dto/ChatRoomDto.kt
package com.example.chatservice.chat.dto

import com.example.chatservice.chat.entity.ChatRoom
import com.example.chatservice.member.dto.MemberDto
import java.io.Serializable

data class ChatRoomDto(
    var id: Long? = null,
    var roomId: String? = null,
    var roomName: String? = null,
    var roomImage: String? = null,
    var lastChatMsg: String? = null,
    var lastChatId: String? = null,
    var lastChatTime: String? = null,
    var members: MutableList<MemberDto> = ArrayList(),
) : Serializable {

    constructor(chatRoom: ChatRoom, members: List<MemberDto>) : this(
        id = chatRoom.id,
        roomId = chatRoom.roomId,
        roomName = chatRoom.roomName,
        roomImage = chatRoom.roomImage,
        lastChatMsg = chatRoom.lastChatMsg,
        lastChatId = chatRoom.lastChatId,
        lastChatTime = chatRoom.lastChatTime,
        members = members.toMutableList(),
    )

    fun updateLastChat(chatId: String?, chatDto: ChatDto) {
        lastChatId = chatId
        lastChatMsg = if (chatDto.chatType == ChatType.TEXT.chatType) {
            chatDto.message
        } else {
            ChatType.fromChatType(chatDto.chatType)?.chatTypeStr
        }
        lastChatTime = chatDto.chatTime
    }

    fun checkChatRoomMember(userId: String): Boolean = members.none { it.userId == userId }
}
